### PR TITLE
Avoid user does not exist error when detecting schedule actions when the commit author is an external user 

### PR DESCRIPTION
**Pull request by yp05327**
image

When repo is a mirror, and commit author is an external user, then `GetUserByEmail` will return error.

reproduce/test:
- mirror Gitea to your instance
- disable action and enable it again, this will trigger `DetectAndHandleSchedules`

ps: also follow #24706, it only fixed normal runs, not scheduled runs.

### CODE REVIEW

File: services/actions/notifier_helper.go [hunk 1]
```diff
@@ -528,7 +528,11 @@ func DetectAndHandleSchedules(ctx context.Context, repo *repo_model.Repository)
 	// Here we use the commit author as the Doer of the notifyInput
 	commitUser, err := user_model.GetUserByEmail(ctx, commit.Author.Email)
 	if err != nil {
-		return fmt.Errorf("get user by email: %w", err)
+		if user_model.IsErrUserNotExist(err) {
+			commitUser = user_model.NewReplaceUser(commit.Author.Name)
```
Comment (lunny) [inline services/actions/notifier_helper.go]: image
Then the TriggerUserID will become 0 and will be stored into database?

Comment (yp05327) [reply services/actions/notifier_helper.go]: yes, 0 will be stored into database. But maybe action user or ghost user is better.

**Comment by yp05327** [comment]
Thanks for the suggestion from @wolfogre 
We should use repo owner as the user instead of commit author.

File: services/actions/notifier_helper.go [hunk 2]
```diff
@@ -525,12 +525,12 @@ func DetectAndHandleSchedules(ctx context.Context, repo *repo_model.Repository)
...
-	commitUser, err := user_model.GetUserByEmail(ctx, commit.Author.Email)
-	if err != nil {
-		return fmt.Errorf("get user by email: %w", err)
+	// if repo is a mirror, commit author maybe external user,
+	// so we need to use the repo owner as the Doer of the notifyInput
+	if err := repo.LoadOwner(ctx); err != nil {
+		return err
 	}
-	notifyInput := newNotifyInput(repo, commitUser, webhook_module.HookEventSchedule)
+	notifyInput := newNotifyInput(repo, repo.Owner, webhook_module.HookEventSchedule)
```
Comment (lunny) [inline services/actions/notifier_helper.go]: How about use gitea-action bot directly?

Comment (yp05327) [reply services/actions/notifier_helper.go]: This sound better.

**Review by lunny — APPROVED** [review]

**Comment by wolfogre** [comment]
Actually, I still think the repo owner is better than the gitea-action user. It's not a technical issue, but a design one.

"gitea-action" means the operation is do by Gitea Actons, like a commit has been pushed and the pusher is "gitea-action", and it shouldn't trigger another Actions run. So it's weird to say that the actor is "gitea-action" since it should never trigger Actions runs.

Whichever, both of them are better than the commit author.

I'm OK with this, since the actor of a scheduled run is not shown.

image

**Review by wolfogre — APPROVED** [review]

**Comment by yp05327** [comment]
@wolfogre 
What about the permission when running the job? Do they have difference?

**Comment by wolfogre** [comment]
> @wolfogre What about the permission when running the job? Do they have difference?

No, the permission when running the job is not decided by the trigger user.

**Pull request merged**